### PR TITLE
add: bid by currency

### PR DESCRIPTION
**Pull request by ruibinx**
{Summary}

Fixes #{issue}

**Type of change**

* [ ] Bug Fix (non-breaking change which fixes an issue)
* [ ] New Feature (non-breaking change which adds functionality)
* [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update
* [ ] This change requires a storage migration
* [ ] This change requires a benchmark

**Checklist**

* [ ] My code follows the style guidelines of this project
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation
* [ ] My changes generate no new warnings
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] New and existing benchmarks pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream modules
* [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pallets/ad/src/lib.rs [hunk 1]
```diff
@@ -729,7 +762,7 @@ impl<T: Config> Pallet<T> {
     }
 
     fn slot_current_fraction_balance(slot: &SlotMetaOf<T>) -> BalanceOf<T> {
```
Comment (coderfengyun) [inline pallets/ad/src/lib.rs]: slot_current_fraction_balance这个名字看起来已经不表意了

File: pallets/ad/src/lib.rs [hunk 2]
```diff
@@ -986,15 +1020,13 @@ impl<T: Config> Pallet<T> {
 
         let ad_meta = Self::ensure_owned_or_delegated_by_ad_id(did, ad_id)?;
 
-        let nft_meta = Nft::<T>::meta(nft_id).ok_or(Error::<T>::NotMinted)?;
```
Comment (coderfengyun) [inline pallets/ad/src/lib.rs:989]: 对于nft mint的判断为什么要删除呢？

File: pallets/ad/src/lib.rs [hunk 3]
```diff
@@ -986,15 +1020,13 @@ impl<T: Config> Pallet<T> {
...
-        ensure!(nft_meta.minted, Error::<T>::NotMinted);
-
         let created = <frame_system::Pallet<T>>::block_number();
 
         // check account has enough balance
-        let fraction_balance = T::Assets::reducible_balance(nft_meta.token_asset_id, &who, false);
+        let fraction_balance: BalanceOf<T> = AdAsset::<T>::balance(&ad_asset, &who);
```
Comment (coderfengyun) [inline pallets/ad/src/lib.rs]: 对于Asset来说balance与reducible_balance语义一致么？

File: pallets/ad/src/lib.rs [hunk 4]
```diff
@@ -354,6 +361,29 @@ pub mod pallet {
             Ok(())
         }
 
+        #[pallet::weight((0, Pays::No))]
+        pub fn bid_with_currency(
```
Comment (coderfengyun) [inline pallets/ad/src/lib.rs:366]: 需要增加测试用例

File: pallets/ad/src/lib.rs [hunk 5]
```diff
@@ -56,7 +56,14 @@ type HashOf<T> = <<T as frame_system::Config>::Hashing as Hash>::Output;
...
 type MetaOf<T> = types::Metadata<BalanceOf<T>, DidOf<T>, HashOf<T>, HeightOf<T>>;
 type NftOf<T> = <T as parami_nft::Config>::AssetId;
-type SlotMetaOf<T> = types::Slot<HashOf<T>, HeightOf<T>, NftOf<T>, AssetsOf<T>, AccountOf<T>>;
+type SlotMetaOf<T> = types::Slot<
+    HashOf<T>,
+    HeightOf<T>,
+    NftOf<T>,
+    AssetsOf<T>,
+    AccountOf<T>,
+    CurrencyOrAsset<AssetsOf<T>>,
```
Comment (coderfengyun) [inline pallets/ad/src/lib.rs:65]: 需要加migration？

File: pallets/ad/src/lib.rs (hunk 3 above)
Comment (coderfengyun) [reply pallets/ad/src/lib.rs]: 是否可以增加一个“锁定一部分余额，查看两者是否表现一致”的case？

File: pallets/ad/src/types.rs [hunk 6]
```diff
@@ -36,3 +40,48 @@ pub struct RewardInfo<Balance> {
...
     pub fungibles: Balance,
 }
+
+#[derive(Clone, Decode, Encode, Eq, PartialEq, RuntimeDebug, MaxEncodedLen, TypeInfo)]
+pub enum CurrencyOrAsset<AssetOf> {
+    Currency,
+    Asset(AssetOf),
+}
+
+pub trait AdAsset<T: Config> {
```
Comment (coderfengyun) [inline pallets/ad/src/types.rs:50]: 是否加一个reducibleBalance？

File: pallets/ad/src/lib.rs (hunk 2 above)
Comment (ruibinx) [reply pallets/ad/src/lib.rs:989]: 这个只有Mint之后才有，移到了bid_with_fraction里

File: pallets/ad/src/lib.rs (hunk 3 above)
Comment (ruibinx) [reply pallets/ad/src/lib.rs]: 这个AdAsset是我自己实现的，里面调用的还是reducible_balance。我把trait的名字改一下。

> 对于Asset来说balance与reducible_balance语义一致么？

翻了一下代码，这两个是不一样的，reducible_balance额外处理了frozen和keepAlive的逻辑。如果没有frozen和keepAlive。

对我们来说应该只需要reducible_balance

File: pallets/ad/src/types.rs (hunk 6 above)
Comment (ruibinx) [reply pallets/ad/src/types.rs:50]: 把balance改成reducibleBalance了

File: pallets/ad/src/migrations.rs [hunk 7]
```diff
@@ -1 +1,97 @@
...
+
+        #[cfg(feature = "try-runtime")]
+        fn post_upgrade() -> Result<(), &'static str> {
+            use log::info;
+            let storage_version = StorageVersion::get::<Pallet<T>>();
+            assert!(storage_version == 5, "current storage version should be 5");
+
+            let mut counter = 0;
+            for (_, slot) in SlotOf::<T>::iter() {
+                info!("new slot {:?} ", slot);
```
Comment (coderfengyun) [inline pallets/ad/src/migrations.rs:94]: 这里是否验证一下SlotAsset已经正确?

File: pallets/ad/src/tests.rs [hunk 8]
```diff
@@ -1402,3 +1417,148 @@ fn should_fail_if_payout_base_too_low() {
...
+        // ensure: deadline, slot, remain
+
+        assert_eq!(<EndtimeOf<Test>>::get(&ad2), Some(1));
+        assert_eq!(<DeadlineOf<Test>>::get(nft, &ad1), None);
+        assert_eq!(<DeadlineOf<Test>>::get(nft, &ad2), Some(1));
+    });
+}
+
+#[test]
+fn should_fail_bid_with_currency_if_minted() {
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:1531]: 需要再加一个should_fail_bid_with_fraction_if_not_minted ?

Comment (ruibinx) [reply pallets/ad/src/tests.rs:1531]: done

File: pallets/ad/src/migrations.rs (hunk 7 above)
Comment (ruibinx) [reply pallets/ad/src/migrations.rs:94]: 线上的slot现在只有一个，人工比一下就行

**Pull request merged**